### ISSUE TITLE
Open API links support

### ISSUE THREAD

**Issue by Stranger6667** [kiwicom/schemathesis#548]
As proposed by @barrett-schonefeld in #458 supporting [Open API links](https://swagger.io/docs/specification/links/) might dramatically improve the Schemathesis testing depth.

Currently, we test endpoints in isolation and there are a lot of 404 responses because there is no information about what actually exists on the server-side.

With Open API links, we can connect the testing of different endpoints. For example, instead of random order and throwing responses away, we can do something like this:
- Create an object
- Retrieve this object by id returned in "Create object"
- Update this object
- Delete this object

It gives a completely different coverage surface!

Regarding the implementation my thoughts are:
- A list of endpoints may be treated a directed graph;
- We can apply topological sorting to the graph's vertices so we have sequences like the mentioned above;
- For the threaded runner, we may need to put the whole group into a single worker, so it is executed sequentially or make some locking/priority mechanism to execute tests with certain order;
- we still need to preserve default tests for "Retrieve" / "Update" / "Delete" parts and do not use data from the linked responses. Otherwise, we will always have some data generated by the server and no data that is actually generated by Hypothesis
- For the process itself, we need to store responses in some kind of storage and this storage should be accessible between tests
- Not sure how to display new tests in the CLI output - maybe add a suffix to the output line (e.g. `GET /users/{user_id} [linked]`)
- Probably this feature should be enabled by default and can be optionally disabled
- Open API 2 can have support for this via a vendor extension

**Comment by barrett-schonefeld** [comment]
For the threaded runner:

It is also an option to build the dependency graphs of the requests first. Then the maximum number of worker threads would be equal to the number of connected dependency graphs.

This would be a simple solution and would work great in this case:

    Thread 1: A -> B -> C
    Thread 2: X -> Y -> Z

It would be less than optimal when a request has many children because the child requests would have to wait on each other even though the information needed for the child request is already available:

    Thread 1: 
                -> B
                -> C
            A   -> D
                -> E
                -> F
                ...

**Comment by Stranger6667** [comment]
@barrett-schonefeld 
Interesting idea! Thanks!

**Comment by Stranger6667** [comment]
I will try to sketch something in the next few days and will share the results :)

**Comment by danhudlow** [comment]
I won't claim that I fully understand the approach for OpenAPI link support being proposed here, but I wonder if it might be overlooking some important scenarios.

Specifically, if `operation_a` and `operation_b` both return some piece of output that's needed as input for `operation_c` it still may not be possible to get construct a successful `operation_c` request if the approach is limited to building a tree outward from a single operation and following its links, as is suggested by the potential CLI output in the PR for this issue:

```
POST /api/users/ . 
  -> GET /api/users/{user_id} .
  -> PUT /api/users/{user_id} .
       -> DELETE /api/users/{user_id} .
```

I wonder if it would be useful to essentially reverse the indexing of links such that for each operation, you know what other operations can provide input. From there, you would still need some strategy to wait for output from operations which link to the one being considered.

It's important to note that cycles could very easily exist in links. Usually these shouldn't be cycles of **required** properties, but even that is possible. (I can provide some detailed scenarios if it would be helpful.)

This might sound a little contrived, but needing output from several requests as input for a request is an incredibly common scenario for APIs I've worked on.

**Comment by Stranger6667** [comment]
> This might sound a little contrived, but needing output from several requests as input for a request is an incredibly common scenario for APIs I've worked on.

It does sound like a valid workflow for me indeed! One of the possible problems I can see at first glance is that there are no rules defined in the spec to resolve priorities when merging two or more computed links into one. It probably can be avoided on the schema level by specifying distinct output parameters/requestBody, but in a general case, there is no such restriction. However, we can do that and set our own merge rules - e.g. just override things in the order they are recorded.

For example:
- `getLocale` operation returns `localeId` for `getUserById`
- `createUser` operation returns `user_id` for `getUserById`
- `updateUser` operation returns `user_id` for `getUserById`
- then we need to call `getUserById` with `localeId` and `user_id`. Which one to use? or produce combinations to check how newly created users are displayed and how updated users are displayed?

I do like the idea, but I feel that before having this we need to define rules and check possible conflicting situations. It will probably broaden the testing surface even more than I expected, but having the current PR size (it is not yet finished) I think that it is better to defer this improvement.

And actually there are some more things that can be improved on our side - for example a more flexible `requestBody` processing. Now the spec defines only literals or expressions for it. I.e. it is not possible to set a particular property, only the whole body. For example, we can deviate from the spec and treat all string as expressions + merge this data with the generated one recursively.

I'll create a separate issue for it :) Thank you for bringing this up @danhudlow

**Comment by Stranger6667** [comment]
Done in #581. An additional issue for the suggested multiple operations feature is #604
